pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -130,6 +135,53 @@ func (nim *nodeInfoManager) InstallCSIDriver(driverName string, driverNodeID str
 	return nil
 }
 
+func (nim *nodeInfoManager) UpdateCSINodeAllocatable(driverName string, newCount int32) error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.MutableCSINodeAllocatableCount) {
+		return nil
+	}
+
+	csiKubeClient := nim.volumeHost.GetKubeClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	newAllocatableCount := &storagev1.VolumeNodeResources{
+		Count: &newCount,
+	}
+
+	err := wait.ExponentialBackoff(updateBackoff, func() (bool, error) {
+		if err := nim.tryUpdateCSINodeAllocatable(csiKubeClient, driverName, newAllocatableCount); err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return fmt.Errorf("error updating CSINode allocatable: %w", err)
+	}
+	return nil
+}
+
+func (nim *nodeInfoManager) tryUpdateCSINodeAllocatable(
+	csiKubeClient clientset.Interface,
+	driverName string,
+	allocatable *storagev1.VolumeNodeResources) error {
+
+	nodeInfo, err := csiKubeClient.StorageV1().CSINodes().Get(context.TODO(), string(nim.nodeName), metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	for i, driver := range nodeInfo.Spec.Drivers {
+		if driver.Name == driverName {
+			nodeInfo.Spec.Drivers[i].Allocatable = allocatable
+			_, err := csiKubeClient.StorageV1().CSINodes().Update(context.TODO(), nodeInfo, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	return fmt.Errorf("driver %s not found in CSINode %s", driverName, nim.nodeName)
```

Comment:
if it's not found, is it worth retrying?

Reply:
This is a good point, if driver not found then there's no update needed so it makes more sense to treat it as a non-error, updated logic.